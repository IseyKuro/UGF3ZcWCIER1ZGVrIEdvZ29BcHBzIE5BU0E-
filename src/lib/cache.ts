import NodeCache from 'node-cache';

const cache = new NodeCache();

export const set = (key, obj) => cache.set(key, obj);

export const get = <T>(key) => cache.get<T>(key);
