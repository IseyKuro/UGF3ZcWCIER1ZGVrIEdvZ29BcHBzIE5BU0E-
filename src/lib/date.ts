import dayjs from "dayjs";
import utc from 'dayjs/plugin/utc';
import timezone from 'dayjs/plugin/timezone';
dayjs.extend(utc)
dayjs.extend(timezone)
dayjs.tz.setDefault('Etc/UTC');

export const differenceInDays = (startDate: string, endDate: string): number => {
  return dayjs(endDate).diff(startDate, 'days');
};

export const divideDateRangeToPeriods = (
  startDate: string,
  endDate: string,
  numOfPeriods: number,
): string[][] => { // FIXME typing
  const periodDays = differenceInDays(startDate, endDate) / numOfPeriods;
  const result = Array.from({ length: numOfPeriods }).reduce<string[][]>((accumulator, _: never, i: number) => {
    const startPeriodDate = (i === 0 
      ? dayjs(startDate) 
      : dayjs(accumulator[accumulator.length - 1][1]).add(1, 'days'));
    const endPeriodDate = i === numOfPeriods - 1 
      ? endDate 
      : startPeriodDate.add(periodDays - 1, 'days').format('YYYY-MM-DD');
      
    return [...accumulator, [startPeriodDate.format('YYYY-MM-DD'), endPeriodDate]]
  }, []);
  console.log(result); // FIXME remove console.log
  return result;
};
