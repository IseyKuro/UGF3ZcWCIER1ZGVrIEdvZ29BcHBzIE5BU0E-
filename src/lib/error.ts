
export type ValidationContext = {
  value: string,
  msg: string,
  param: string,
  location: string,
}
export interface HTTPError extends Error {
  code: number;
}
export interface HTTPValidationError<ContextType> extends HTTPError {
  context: ContextType;
}

export class HTTPError extends Error {
  constructor(code: number, message?: string) {
    super(message);
    this.name = 'Error';
    this.code = code;
  }

  toMessage() {
    return { ...this };
  }
}

export class HTTPValidationError<ContextType> extends HTTPError {
  constructor(context: ContextType) {
    super(400);
    this.name = 'Validation Error';
    this.context = context;
  }
}

export class HTTPInternalError extends HTTPError {
  constructor() {
    super(500);
    this.name = 'Internal Server Error';
  }
}
