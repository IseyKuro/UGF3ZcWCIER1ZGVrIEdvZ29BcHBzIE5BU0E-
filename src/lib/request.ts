import https from 'https';

export const request = (url: string): Promise<JSON | Error> => new Promise((resolve, reject) => {
  https.get(url, (res) => {
    const data = [];
  
    res.on('data', (chunk) => {
      data.push(chunk as never);
    });
  
    res.on('end', () => {
      resolve(JSON.parse(Buffer.concat(data).toString()));
    });
  }).on('error', (error) => {
    reject(error);
  })
});
