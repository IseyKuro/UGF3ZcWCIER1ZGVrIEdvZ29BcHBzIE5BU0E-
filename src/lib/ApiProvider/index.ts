import { HTTPError, HTTPInternalError } from "../error";
import { getRequest, GetRequest, RequestResponse } from "./getRequest";

export type ObjectType = Record<string, string>;
export type Response = unknown | Error;
export type HandleResponseInput = JSON | Error;
export type HandleResponse<T> = (response: any) => T | Error;
export type HandleError = (error: Error) => HTTPError;
export type BuildUrl<P> = (params: P, daysBetweenPeriod: number) => string;
export type CalculateRequestConcurency<P> = (params: P) => number;
export type DivideQueryParams<P> = (params: P, concurency: number) => P[];

export class ApiProvider<P, T> {
  private MAX_CONCURENT_REQUESTS;
  private handleResponse;
  private buildUrl;
  private handleError;
  private calculateRequestConcurency;
  private divideQueryParams;
  private request = getRequest;
  constructor(
    maxConcurrentRequests: number,
    handleResponse: HandleResponse<T>,
    handleError: HandleError,
    buildUrl: BuildUrl<P>,
    calculateRequestConcurency: CalculateRequestConcurency<P>,
    divideQueryParams: DivideQueryParams<P>,
    request?: GetRequest,
  ) {
    this.MAX_CONCURENT_REQUESTS = maxConcurrentRequests;
    this.handleResponse = handleResponse;
    this.handleError = handleError;
    this.buildUrl = buildUrl;
    this.calculateRequestConcurency = calculateRequestConcurency;
    this.divideQueryParams = divideQueryParams;
    if (request) {
      this.request = request;
    }
  }

  public async call(params: ObjectType): Promise<T | T[] | undefined> {
    const concurency = this.calculateRequestConcurency(params);
    if (concurency > Number(this.MAX_CONCURENT_REQUESTS)) {
      throw new HTTPInternalError();
    }
    try {
      return concurency === 1
        ? await this.getData(params)
        : (await Promise.allSettled(this.divideQueryParams(params, concurency)
          .map((dividedParams) => this.getData(dividedParams))) as unknown as T[]); // FIXME typing
    } catch (error) {
      this.handleError(error);
    }
  }

  private async getData(params: ObjectType): Promise<T> {
    const url = this.buildUrl(params);
    const response = await this.request(url);
    return this.handleResponse(response);
  }
}
