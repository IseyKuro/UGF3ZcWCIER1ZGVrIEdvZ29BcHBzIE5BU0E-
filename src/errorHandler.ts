import { ErrorRequestHandler, NextFunction, Request, Response } from "express";
import { HTTPError } from "./lib/error";

export const httpErrorHandler = (
  error: Error,
  req: Request,
  res: Response,
  next: NextFunction,
): void => {
  if (error instanceof HTTPError) {
    res.status(error.code).json(error.toMessage()).send();
  } else {
    next(error);
  }
};

export const unknownErrorHandler = (
  error: Error,
  req: Request,
  res: Response,
  _: NextFunction,
): void => {
  console.error('Unknow error occured! ', error.message);
  res.status(500).json({ error: 'Internal Error' }).send();
};

export const errorHandlers: ErrorRequestHandler[] = [httpErrorHandler, unknownErrorHandler];
