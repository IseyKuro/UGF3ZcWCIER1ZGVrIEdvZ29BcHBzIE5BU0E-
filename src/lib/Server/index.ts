import express, { Router, Express, ErrorRequestHandler, RequestHandler } from 'express';
import { errorHandler } from './errorHandler';
import queryString from 'query-string';

export class Server {
  private port = '8080';
  private app: Express;
  private timeout = 120000;
  constructor(timeout?: number) {
    this.timeout = timeout || this.timeout;
    this.app = express();
    this.app.use(...errorHandler);
  }

  public static create(timeout?: number): Server {
    return new Server(timeout);
  }

  public registerErrorHandlers(errorHandlers: ErrorRequestHandler[]) {
    this.app.use(...errorHandlers);
    return this;
  }

  public addRoutes(routes: Router[]): Server {
    routes.forEach((route) => this.app.use(route));
    return this;
  }

  public registerMiddleware(fn: RequestHandler): Server {
    this.app.use(fn);
    return this;
  }

  public enableQueryParser(): Server {
    this.app.set('query parser', (str) => queryString.parse(str));
    return this;
  }

  public setPort(port: string): Server {
    this.port = port;
    return this;
  }

  public init() {
    const server = this.app.listen(this.port, () => {
      console.info(`Server listening on port: ${this.port}`);
    });
    server.setTimeout(this.timeout);

    return server;
  }
}
