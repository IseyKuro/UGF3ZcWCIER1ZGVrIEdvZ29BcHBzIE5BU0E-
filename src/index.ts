import * as dotenv from 'dotenv';
dotenv.config();
import { NextFunction, Request, Response } from 'express';
import { Server } from './lib/Server/index';
import nasaRouter from './nasa/nasaRouter';
import { init as concurencyServiceInit } from './nasa/concurencyService';
import { errorHandlers } from './errorHandler';

const {
  SERVER_PORT,
  SERVER_TIMEOUT
} = process.env;

export const application = Server
  .create()
  .enableQueryParser()
  .registerEntryMiddleware((req: Request, _res: Response, next: NextFunction) => {
    const now = new Date().toISOString();
    console.log(`${now} Registered new request from ip: ${req.ip}`);
    next();
  })
  .addRoutes([nasaRouter])
  .registerErrorHandlers(errorHandlers)
  .build();

concurencyServiceInit().then(() => {
  application.listen(SERVER_PORT, () => {
    console.info(`Server listening on port: ${SERVER_PORT}`);
  }).setTimeout(parseInt(SERVER_TIMEOUT || ''));
});
