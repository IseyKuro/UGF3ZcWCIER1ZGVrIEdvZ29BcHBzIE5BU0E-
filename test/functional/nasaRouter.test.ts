import request from 'supertest';
import { application } from '../../src';
import { request as getRequest } from '../../src/lib/request';

jest.mock('../../src/lib/request');

describe('Nasa router test', () => {
  let app;

  beforeAll(() => {
    app = request.agent(application.listen(process.env.PORT));
  });

  it('Should response with status code 400', async () => {
    const response = await app.get('/').send();

    expect(response.status).toBe(400);
    expect(response.body).toMatchSnapshot();
  });

  it('Should response with status code 200', async () => {
    (getRequest as jest.Mock).mockResolvedValueOnce([{
      hdurl: "https://apod.nasa.gov/apod/image/1901/sombrero_spitzer_3000.jpg",
    }]);

    const response = await app
      .get('/?startDate=2019-01-01&endDate=2019-01-01')
      .send();

    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot();
  });

  it('Should response with status code 503', async () => {
    (getRequest as jest.Mock).mockImplementation(async () => new Promise((res) => {
      setTimeout(() => {
        res([{ hdurl: "https://apod.nasa.gov/apod/image/1901/sombrero_spitzer_3000.jpg"} ])
      }, 0)
    }));

    const response = await app
      .get('/?startDate=2010-01-01&endDate=2019-01-01')
      .send();
    
    expect(response.status).toBe(503);
    expect(response.body).toMatchSnapshot();
  });

  it('Should response with status code 503', async () => {
    (getRequest as jest.Mock).mockImplementation(async () => new Promise((res) => {
      setTimeout(() => {
        res([{ hdurl: "https://apod.nasa.gov/apod/image/1901/sombrero_spitzer_3000.jpg"} ])
      }, 1000)
    }));

    const response = await Promise.all([
      app.get('/?startDate=2018-01-01&endDate=2019-01-01').send(),
      app.get('/?startDate=2017-01-01&endDate=2018-01-01').send(),
      app.get('/?startDate=2016-01-01&endDate=2017-01-01').send(),
      app.get('/?startDate=2015-01-01&endDate=2016-01-01').send(),
      app.get('/?startDate=2014-01-01&endDate=2015-01-01').send(),
      app.get('/?startDate=2013-01-01&endDate=2014-01-01').send(),
    ]);
    
    expect(response).toBeDefined();
    expect(response[5].status).toBe(503);
    expect(response[5].body).toMatchSnapshot();
  });
});
