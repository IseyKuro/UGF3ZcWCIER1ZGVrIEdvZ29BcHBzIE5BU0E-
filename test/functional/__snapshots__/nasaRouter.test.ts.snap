// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nasa router test Should response with status code 200 1`] = `
{
  "urls": [
    "https://apod.nasa.gov/apod/image/1901/sombrero_spitzer_3000.jpg",
  ],
}
`;

exports[`Nasa router test Should response with status code 400 1`] = `
{
  "code": 400,
  "context": [
    {
      "instancePath": "",
      "keyword": "required",
      "message": "must have required property 'startDate'",
      "params": {
        "missingProperty": "startDate",
      },
      "schemaPath": "#/required",
    },
  ],
  "error": "Validation Error",
}
`;

exports[`Nasa router test Should response with status code 503 1`] = `
{
  "code": 503,
  "error": "Unavailable Error",
}
`;

exports[`Nasa router test Should response with status code 503 2`] = `
{
  "code": 503,
  "error": "Unavailable Error",
}
`;
